clients/vault/src/oracle/types.rs

```diff
@@ -22,47 +26,103 @@ pub(crate) type SlotEncodedMap = BTreeMap<Slot, SerializedData>;
 
 pub(crate) type SlotList = BTreeMap<Slot, ()>;
 
-/// This is used to add some helpers that make using the VecDeque easier.
-pub trait LifoMap<T, K> {
-	fn get_with_key(&self, key: &K) -> Option<&T>;
-
-	fn set_with_key(&mut self, key: K, value: T);
-
-	fn remove_with_key(&mut self, key: &K) -> Option<T>;
-
-	fn contains_key(&self, key: &K) -> bool;
+#[derive(Debug, Clone)]
+pub struct LimitedFifoMap<K, T> {
+	limit: usize,
+	queue: VecDeque<(K, T)>,
 }
 
-impl<T, K> LifoMap<T, K> for VecDeque<(K, T)>
+impl<K, T> LimitedFifoMap<K, T>
 where
-	K: PartialEq,
+	K: Debug + PartialEq,
 {
-	fn get_with_key(&self, key: &K) -> Option<&T> {
-		self.iter().find(|(k, _)| k == key).map(|(_, v)| v)
+	pub fn new() -> Self {
+		LimitedFifoMap { limit: DEFAULT_MAX_ITEMS_IN_QUEUE, queue: VecDeque::new() }
+	}
+
+	pub fn with_limit(mut self, limit: usize) -> Self {
+		// cannot set a number smaller than the default.
+		if limit < DEFAULT_MAX_ITEMS_IN_QUEUE {
+			self.limit = DEFAULT_MAX_ITEMS_IN_QUEUE
+		} else {
+			self.limit = limit;
+		}
+
+		self
+	}
+
+	pub fn limit(&self) -> usize {
+		self.limit
+	}
+
+	pub fn len(&self) -> usize {
+		self.queue.len()
+	}
+
+	pub fn contains(&self, key: &K) -> bool {
+		self.queue.iter().any(|(k, _)| k == key)
 	}
 
-	fn set_with_key(&mut self, key: K, value: T) {
-		// If the key already exists, remove it first
-		self.remove_with_key(&key);
+	pub fn get(&self, key: &K) -> Option<&T> {
+		self.queue.iter().find(|(k, _)| k == key).map(|(_, v)| v)
+	}
+
+	pub fn first(&self) -> Option<&(K, T)> {
+		self.queue.get(0)
+	}
 
-		self.push_back((key, value));
+	pub fn iter(&self) -> std::collections::vec_deque::Iter<'_, (K, T)> {
+		self.queue.iter()
 	}
 
-	fn remove_with_key(&mut self, key: &K) -> Option<T> {
-		let index = self.iter().find_position(|(k, _)| k == key);
-		if let Some((index, _)) = index {
-			return self.remove(index).map(|(_, v)| v)
+	pub fn remove(&mut self, key: &K) -> Option<T> {
+		let (index, _) = self.queue.iter().find_position(|(k, _)| k == key)?;
+		self.queue.remove(index).map(|(_, v)| v)
+	}
+
+	pub fn insert(&mut self, key: K, value: T) -> Option<T> {
+		let old_value = self.remove(&key);
+
+		// remove the oldest entry if the queue reached its limit
+		if self.queue.len() == self.limit {
+			if let Some(oldest_entry) = self.queue.pop_front() {
+				tracing::debug!("removing old entry with key: {:?}", oldest_entry.0);
+			}
 		}
-		None
+
+		self.queue.push_back((key, value));
+
+		old_value
+	}
+
+	/// Consumes the other and returns the excess of it. The limit will be based on self.
+	pub fn append(&mut self, other: Self) -> VecDeque<(K, T)> {
+		// check the remaining size available for this map.
+		let allowable_size = self.limit - self.len();
+
+		let mut other_queue = other.queue;
+
+		let last_partition = if allowable_size < other_queue.len() {
+			// split off the 'other' map, since it's too big to append all of its elements
+			other_queue.split_off(allowable_size)
```

Comment:
Interesting approach. Now something that comes to my mind is, whether we want to have the 'younger' or 'older' split of elements when we cut this off. We probably want to have the 'younger' elements, and if I understand correctly your implementation already does exactly that :+1:

Reply:
@ebma  can you review again? 

you can check the test case at lines 274 to 281